File: Pilha/pilha.rs
const MAX: usize = 15; // tamanho máximo da pilha

struct Pilha {
    top: i32,
    items: [i32; MAX],
}

impl Pilha {
    fn new() -> Pilha {
        Pilha {
            top: -1,
            items: [0; MAX],
        }
    }

    fn is_empty(&self) -> bool {
        self.top == -1
    }

    fn is_full(&self) -> bool {
        self.top as usize == MAX - 1
    }

    fn push(&mut self, x: i32) {
        if self.is_full() {
            println!("Erro: a pilha está cheia");
        } else {
            self.top += 1;
            self.items[self.top as usize] = x;
        }
    }

    fn pop(&mut self) -> Option<i32> {
        if self.is_empty() {
            println!("Erro: a pilha está vazia");
            None
        } else {
            let x = self.items[self.top as usize];
            self.top -= 1;
            Some(x)
        }
    }

    fn peek(&self) -> Option<i32> {
        if self.is_empty() {
            println!("Erro: a pilha está vazia");
            None
        } else {
            let x = self.items[self.top as usize];
            Some(x)
        }
    }

    fn display(&self) {
        if self.is_empty() {
            println!("A pilha está vazia");
        } else {
            println!("Itens na pilha:");
            for i in (0..=self.top).rev() {
                println!("{}", self.items[i as usize]);
            }
        }
    }
}

fn main() {
    let mut p = Pilha::new();

    loop {
        println!("Escolha uma opção:");
        println!("1 - Inserir elemento");
        println!("2 - Retirar elemento");
        println!("3 - Ver elemento no topo");
        println!("4 - Ver pilha toda");
        println!("0 - Sair");

        let mut opcao = String::new();
        std::io::stdin().read_line(&mut opcao).expect("Erro ao ler a opção");

        match opcao.trim().parse::<i32>() {
            Ok(0) => break,
            Ok(1) => {
                println!("Digite o elemento a ser inserido:");
                let mut elem = String::new();
                std::io::stdin().read_line(&mut elem).expect("Erro ao ler o elemento");
                let x = elem.trim().parse::<i32>().expect("Erro ao converter o elemento");
                p.push(x);
            }
            Ok(2) => {
                match p.pop() {
                    Some(x) => println!("Elemento retirado: {}", x),
                    None => (),
                }
            }
            Ok(3) => {
                match p.peek() {
                    Some(x) => println!("Elemento no topo: {}", x),
                    None => (),
                }
            }
            Ok(4) => p.display(),
            _ => println!("Opção inválida"),
        }
    }
}

File: Pilha/pilha.py
MAX_SIZE = 100

# Definição da classe para a pilha
class Pilha:
    def __init__(self):
        self.dados = [None] * MAX_SIZE # Lista para armazenar os dados
        self.topo = -1 # Variável para rastrear o topo da pilha

    # Função para verificar se a pilha está vazia
    def is_vazia(self):
        return self.topo == -1 # Retorna True se o topo for igual a -1, indicando pilha vazia

    # Função para verificar se a pilha está cheia
    def is_cheia(self):
        return self.topo == MAX_SIZE - 1 # Retorna True se o topo for igual ao tamanho máximo da pilha

    # Função para inserir uma string na pilha
    def inserir_string(self, string):
        if self.is_cheia():
            print("Erro: Pilha cheia")
            return
        self.topo += 1 # Incrementa o topo
        self.dados[self.topo] = string # Insere a string no topo da pilha

    # Função para verificar se uma string é um número real (float) ou inteiro (int)
    def is_numero(self, string):
        try:
            float(string)
            return True
        except ValueError:
            return False

    # Função para converter uma string para número real (float) ou inteiro (int)
    def converter_numero(self, string):
        if self.is_numero(string):
            inteiro = int(string) # Converte a string para inteiro
            print("Número inteiro:", inteiro)
        else:
            real = float(string) # Converte a string para float
            print("Número real: {:.2f}".format(real))

    # Função para remover o topo da pilha
    def remover_topo(self):
        if self.is_vazia():
            print("Erro: Pilha vazia")
            return
        self.topo -= 1 # Decrementa o topo

    # Função para verificar o topo da pilha
    def ver_topo(self):
        if self.is_vazia():
            print("Erro: Pilha vazia")
            return
        print("Topo da pilha:", self.dados[self.topo])

    # Função para visualizar toda a pilha
    def ver_pilha(self):
        if self.is_vazia():
            print("Pilha vazia")
            return
        print("Conteúdo da pilha:")
        for i in range(self.topo, -1, -1):
            print(self.dados[i])


pilha = Pilha()
while True:
    print("\nEscolha uma opção:")
    print("0. Sair")
    print("1. Inserir string")
    print("2. Verificar se é número e converter")
    print("3. Remover topo")
    print("4. Ver topo")
    print("5. Ver pilha toda")

    opcao = int(input("Opção: "))

    if opcao == 0:
        print("Saindo...")
        break
    elif opcao == 1:
        entrada = input("Digite a string a ser inserida: ")
        pilha.inserir_string(entrada)
    elif opcao == 2:
        entrada = input("Digite a string a ser verificada: ")
        pilha.converter_numero(entrada)
    elif opcao == 3:
        pilha.remover_topo()
    elif opcao == 4:
        pilha.ver_topo()
    else:
        pilha.ver_pilha()
     
